base/applications/rapps/geninst.cpp

```diff
@@ -0,0 +1,813 @@
+    SendMessage(Info.hDlg, IM_PROGRESS, 0, 0); // Cleanup
+    return 0;
+}
+
+static DWORD CALLBACK
+UIWorkerThread(LPVOID Parameter)
+{
+    ((LPTHREAD_START_ROUTINE)Parameter)(NULL);
+    return SendMessage(g_pInfo->hDlg, IM_END, 0, 0);
+}
+
+static INT_PTR CALLBACK
+UIDlgProc(HWND hDlg, UINT uMsg, WPARAM wParam, LPARAM lParam)
+{
+    HWND hPB = GetDlgItem(hDlg, 1);
+    switch(uMsg)
+    {
+        case IM_STARTPROGRESS:
+            SetWindowLongPtr(hPB, GWL_STYLE, WS_CHILD | WS_VISIBLE);
+            SendMessageW(hPB, PBM_SETMARQUEE, FALSE, 0);
+            SendMessageW(hPB, PBM_SETRANGE32, 0, g_pInfo->Count);
+            SendMessageW(hPB, PBM_SETPOS, 0, 0);
+            break;
+        case IM_PROGRESS:
+            SendMessageW(hPB, PBM_DELTAPOS, 1, 0);
+            break;
+        case IM_END:
+            DestroyWindow(hDlg);
+            break;
+        case WM_INITDIALOG:
+        {
+            SendMessageW(hPB, PBM_SETMARQUEE, TRUE, 0);
+            g_pInfo->hDlg = hDlg;
+            HICON hIco = LoadIconW(hInst, MAKEINTRESOURCEW(IDI_MAIN));
+            SendMessageW(hDlg, WM_SETICON, ICON_BIG, (LPARAM)hIco);
+            SendMessageW(hDlg, WM_SETTEXT, 0, (LPARAM)g_pInfo->AppName);
+            if (!SHCreateThread(UIWorkerThread, (void*)lParam, CTF_COINIT, NULL))
+            {
+                ErrorBox();
+                SendMessageW(hDlg, IM_END, 0, 0);
+            }
+            break;
+        }
+        case WM_CLOSE:
+            return TRUE;
+        case WM_DESTROY:
+            PostMessage(NULL, WM_QUIT, 0, 0);
+            break;
+    }
+    return FALSE;
+}
+
+static BOOL
+CreateUI(BOOL Silent, LPTHREAD_START_ROUTINE ThreadProc)
+{
+    enum { DLGW = 150, DLGH = 20, PAD = 4, PADx2 = PAD * 2, CHILDCOUNT = 1 };
+    const UINT DlgStyle = WS_CAPTION | DS_MODALFRAME | DS_NOFAILCREATE | DS_CENTER;
+    static const WORD DlgTmpl[] =
+    {
+        LOWORD(DlgStyle), HIWORD(DlgStyle), 0, 0, CHILDCOUNT, 0, 0, DLGW, DLGH, 0, 0, 0,
+        PBS_MARQUEE, HIWORD(WS_CHILD | WS_VISIBLE), 0, 0, PAD, PAD, DLGW - PADx2, DLGH - PADx2, 1,
+        'm', 's', 'c', 't', 'l', 's', '_', 'p', 'r', 'o', 'g', 'r', 'e', 's', 's', '3', '2', 0, 0,
+    };
+    HWND hWnd = CreateDialogIndirectParamW(NULL, (LPCDLGTEMPLATE)DlgTmpl, NULL,
+                                           UIDlgProc, (LPARAM)ThreadProc);
+    if (!hWnd)
+    {
+        ErrorBox();
+        return FALSE;
+    }
+    else if (!Silent)
+    {
+        ShowWindow(hWnd, SW_SHOW);
+    }
+    MSG Msg;
+    while (GetMessageW(&Msg, NULL, 0, 0))
+    {
+        TranslateMessage(&Msg);
+        DispatchMessageW(&Msg);
+    }
+    return TRUE;
+}
+
+BOOL
+ExtractAndRunGeneratedInstaller(const CAvailableApplicationInfo &AppInfo, LPCWSTR Archive)
+{
+    InstallInfo Info(AppInfo.szDisplayName, *AppInfo.GetConfigParser(), Archive);
+    g_pInfo = &Info;
+    return CreateUI(Info.Silent, ExtractAndInstallThread) ? !Info.Error : FALSE;
+}
+
+struct UninstallInfo : CommonInfo
```

Comment:
I am not a big fan of defining structs in the middle of a bunch of functions.

Reply:
The file is organized with the installer bits first and then the uninstaller.